### PR TITLE
Script cactus

### PR DESCRIPTION
**Pull request by dthybert**
## Description

This is a standalone script that dump genomes from core db. The genome can be dumped either as unmasked, soft masked and hard masked.

**Related JIRA tickets:**
- ENSCOMPARASW-4612

## Overview of changes
The script has been written from scratch. Some code have been borrowed from Bio::EnsEMBL::Compara::Production::EPOanchors::DumpGenomeSequence; 

#### Change 1
The script is getting all toplevel slices form a core db and then write each sequence in to a fasta file using a Bio::EnsEMBL::Utils::IO::FASTASerializer object. The sequence can be unmasked, soft masked or hard masked depending on what the user is providing in parameters.

## Testing
Test that the unmasked sequence is dumped when not using the option -mask
Test that the soft masked sequence is dumped when using the option -mask soft
Test that the hard masked sequence is dumped when using the option -mask hard
Test the help message

## Notes
I would like to know if the comments in the script are done as expected.

### CODE REVIEW

File: scripts/dumps/dumpGenomeFromCore.pl [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+my $host;
+my $port;
+my $mask;
+my $genome_dump_file;
+my $help;
+my $desc = "
+This script dump all toplevel sequences from a core database and store it at the fatsa.
+The sequences can be unmasked, soft masked or hard masked.
+
+USAGE dumpGenomeFromCore.pl [-mask] -coreDB COREDB -host HOST -port PORT -outfile OUTFILE
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dumpGenomeFromCore.pl:37]: Most of our scripts don't follow a camel case for their name/parameters:
```suggestion
USAGE dump_genome_from_core.pl [-mask] -core-db COREDB -host HOST -port PORT -outfile OUTFILE
```

File: scripts/dumps/dumpGenomeFromCore.pl [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+Options:
+* -coreDB
+      the core database storing the genome sequence to be dumped
+* --host
+      server hosting the core database
+* --port
+      port for the host database
+* --outfile
+      file where the dumped seqeunce will be sored at the fasta format
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dumpGenomeFromCore.pl:47]:
```suggestion
      file where the dumped sequence will be stored in fasta format
```

File: scripts/dumps/dumpGenomeFromCore.pl [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    'mask=s'     => \$mask,
+    'outfile=s'  => \$genome_dump_file,
+    'help' => \$help
+  );
+
+
+if ($help) {
+  print $desc;
+  exit(0);
+}
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dumpGenomeFromCore.pl:66]: For future reference: alternatively, you can fill in the POD and then use `POD::Usage`. Here is an example: https://github.com/Ensembl/ensembl-compara/blob/master/scripts/pipeline/update_genome.pl

File: scripts/dumps/dumpGenomeFromCore.pl [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                                               -port   => $port,
+                                               -driver => 'mysql');
+
+my $slice_adaptor = $dba->get_adaptor("Slice");
+my $slices = $slice_adaptor->fetch_all("toplevel");
+
+open(my $filehandle, '>', $genome_dump_file) or die "can't open $genome_dump_file for writing\n";
+
+
+# create a fasta serialiser that define the seq_region_name() as fasta header
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dumpGenomeFromCore.pl:80]:
```suggestion
# create a fasta serialiser that defines the seq_region_name() as fasta header
```

File: scripts/dumps/dumpGenomeFromCore.pl [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+my $dba = Bio::EnsEMBL::DBSQL::DBAdaptor->new( -user   => 'ensro',
+                                               -dbname => $dbname,
+                                               -host   => $host,
+                                               -port   => $port,
+                                               -driver => 'mysql');
+
+my $slice_adaptor = $dba->get_adaptor("Slice");
+my $slices = $slice_adaptor->fetch_all("toplevel");
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dumpGenomeFromCore.pl:75]: Alternatively:
```suggestion
my $slices = $dba->get_SliceAdaptor->fetch_all("toplevel");
```
Just to make sure, you don't want non-reference slices, duplicates or LRGs, right?

**Review by JAlvarezJarreta — CHANGES_REQUESTED** [review]
Although I have found some cases, I think it is best if all scripts follow the same name format convention. In this case, could you rename it to `dump_genome_from_core.pl` please? Besides that and a few minor comments, nice and simple code :+1:

File: scripts/dumps/dumpGenomeFromCore.pl [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+use Bio::EnsEMBL::DBSQL::DBAdaptor;
+
+use Getopt::Long;
+
+my $dbname;
+my $host;
+my $port;
+my $mask;
+my $genome_dump_file;
+my $help;
```
[the remark this answers is not in the document]
Comment (dthybert) [reply scripts/dumps/dumpGenomeFromCore.pl:32]: ok, I will change this

File: scripts/dumps/dumpGenomeFromCore.pl (hunk 3 above)
Comment (dthybert) [reply scripts/dumps/dumpGenomeFromCore.pl:66]: thank you for the tip.

File: scripts/dumps/dumpGenomeFromCore.pl [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    'outfile=s'  => \$genome_dump_file,
+    'help' => \$help
+  );
+
+
+if ($help) {
+  print $desc;
+  exit(0);
+}
+
```
[the remark this answers is not in the document]
Comment (dthybert) [reply scripts/dumps/dumpGenomeFromCore.pl:67]: good point!

[the remark this answers is not in the document]
Comment (dthybert) [reply scripts/dumps/dumpGenomeFromCore.pl:67]: good point!

File: scripts/dumps/dumpGenomeFromCore.pl (hunk 5 above)
Comment (dthybert) [reply scripts/dumps/dumpGenomeFromCore.pl:75]: yes we just want reference slices

File: scripts/dumps/dumpGenomeFromCore.pl [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+open(my $filehandle, '>', $genome_dump_file) or die "can't open $genome_dump_file for writing\n";
+
+
+# create a fasta serialiser that define the seq_region_name() as fasta header
+my $serializer = Bio::EnsEMBL::Utils::IO::FASTASerializer->new($filehandle,
+		  sub{
+            my $slice = shift;
+            $slice->seq_region_name();
+            return $slice->seq_region_name();
+          });
```
[the remark this answers is not in the document]
Comment (dthybert) [reply scripts/dumps/dumpGenomeFromCore.pl:86]: I agree it is a much better way to format

File: scripts/dumps/dump_genome_from_core.pl [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+my ($dbname, $host, $port, $mask, $genome_dump_file, $help);
+my $desc = "
+This script dumps all toplevel sequences from a core database and stores them in fasta file.
+The sequences can be unmasked, soft masked or hard masked.
+
+USAGE dump_genome_from_core.pl [-mask] -core-db COREDB -host HOST -port PORT -outfile OUTFILE
+
+Options:
+* -core-db
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dump_genome_from_core.pl:35]: To follow the same format as the rest:
```suggestion
* --core-db
```

File: scripts/dumps/dump_genome_from_core.pl [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      the core database name storing the genome sequence to be dumped
+* --host
+      server hosting the core database
+* --port
+      port for the host database
+* --outfile
+      file where the dumped sequence will be sored in fasta format
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dump_genome_from_core.pl:42]:
```suggestion
      file where the dumped sequence will be stored in fasta format
```

File: scripts/dumps/dump_genome_from_core.pl [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+      sequence will be unmasked.
+";
+
+GetOptions(
+    'coreDB=s'  => \$dbname,
+    'host=s'    => \$host,
+    'port=s'    => \$port,
+    'mask=s'    => \$mask,
+    'outfile=s' => \$genome_dump_file,
+    'help'      => \$help
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dump_genome_from_core.pl:54]: Change in argument name and corrected indentation:
```suggestion
    'core-db|core_db=s' => \$dbname,
    'host=s'            => \$host,
    'port=s'            => \$port,
    'mask=s'            => \$mask,
    'outfile=s'         => \$genome_dump_file,
    'help'              => \$help
```

File: scripts/dumps/dump_genome_from_core.pl [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+  );
+
+
+if ($help) {
+  print $desc;
+  exit(0);
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dump_genome_from_core.pl:60]: Wrong indentation:
```suggestion
    print $desc;
    exit(0);
```

File: scripts/dumps/dump_genome_from_core.pl [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+                                               -host   => $host,
+                                               -port   => $port,
+                                               -driver => 'mysql');
+
+my $slices = $dba->get_SliceAdaptor->fetch_all("toplevel");
+
+open(my $filehandle, '>', $genome_dump_file) or die "can't open $genome_dump_file for writing\n";
+
+
+# create a fasta serialiser that define the seq_region_name() as fasta header
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dump_genome_from_core.pl:78]:
```suggestion
# create a fasta serialiser that defines the seq_region_name() as fasta header
```

File: scripts/dumps/dump_genome_from_core.pl [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+my $serializer = Bio::EnsEMBL::Utils::IO::FASTASerializer->new(
+    $filehandle,
+    sub{
+        my $slice = shift;
+        $slice->seq_region_name();
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dump_genome_from_core.pl:83]: Unnecessary:
```suggestion
```

File: scripts/dumps/dump_genome_from_core.pl [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+foreach my $slice (@$slices){
+    if (defined $mask && $mask eq "soft"){
+        $slice = $slice->get_repeatmasked_seq(undef, 1);
+    }
+    elsif (defined $mask && $mask eq "hard"){
+        $slice = $slice->get_repeatmasked_seq();
+    }
+    my $padded_slice = Bio::EnsEMBL::PaddedSlice->new(-SLICE => $slice);
+    $serializer->print_Seq($padded_slice);
+}
```
Comment (JAlvarezJarreta) [inline scripts/dumps/dump_genome_from_core.pl:98]: End of file should be with a newline:
```suggestion
}

```

**Review by JAlvarezJarreta — CHANGES_REQUESTED** [review]

**Review by JAlvarezJarreta — APPROVED** [review]
Not sure if you have done it already, but before merging, could you confirm the new code to check if the mask is either `soft` or `hard` only works as expected? Once that is done, feel free to merge :+1:

**Comment by dthybert** [comment]
yes I did this was one of my first tests

**Pull request merged**